### DISCUSSION TITLE
How should extensions connect?

### DISCUSSION THREAD

**Discussion by Blatman** [neutralinojs/neutralinojs#737] [Q&A] [answered]
May have a bug here ....

Used getsStats in main.js and got  -

stats: - {connected: [], loaded: ["js.neutralino.sampleextension", "js.neutralino.driverprocess"]} 

Seems to indicate that extns have loaded as per the config file but not connected.

main.js file -

// get stats of connected extns
let stats = Neutralino.extensions.getStats()
.then(function(stats) {
    console.log('stats: ', stats);
});

Log result is --> stats: - {connected: [], loaded: ["js.neutralino.sampleextension", "js.neutralino.driverprocess"]}

// This request will be queued and processed when the extension connects.
//var output = Neutralino.extensions.dispatch("js.neutralino.sampleextension", "eventToExtension", "Hello extension!")

var extnId = "js.neutralino.sampleextension";
var extnEvent = "eventToExtension";
var data = {
    myData: "Hello extension!"
};

Neutralino.extensions.dispatch(extnId, extnEvent, data)
.catch((err) => {
    console.log("Extension isn't loaded! - " + JSON.stringify(err));
});

Dispatch command result -->  Extns have loaded but main.js in the sampleextension folder is not being run.

Config file -

    "extensions": [
        {
            "id": "js.neutralino.sampleextension",
            "command": "node ${NL_PATH}/extensions/sampleextension/main.js"
        },
        {
            "id": "js.neutralino.driverprocess"
        }
    ]

the NL_PATH in my case is -  /Users/USER/Desktop/AppsNeu400/SampleNeu-basic.app/Contents/Resources so command reads as -

node /Users/USER/Desktop/AppsNeu400/SampleNeu-basic.app/Contents/Resources/extensions/sampleextension/main.js

This does not appear to connect but if I run `node main.js` via the bash terminal in sampleextension folder the main.js runs OK.

If I move the terminal to the extensions folder then -   node sampleextensions/main.js runs OK
                                                         node /sampleextensions/main.js fails

Is there an issue with the leading slash on the NL_PATH global variable in this case?

MacOS 10.13
NS: 4.0.0
CL: 3.0.0

**Comment by shalithasuranga** [comment]
Hey.. thanks for trying the new extensions system. Extensions are disabled by default, so you need to enable them via `--enable-extensions` or `enaleExtensions: true` in config.

**Comment by Blatman** [comment]
@shalithasuranga - yes I realized that as when I initially tried I would get extensions not loaded log but when the enableExtensions config was set to true the error log disappeared which suggested that the dispatch command was being sent OK however not getting to main.js - the extensions are loading as per the getStats but just not connecting.

**Reply by shalithasuranga** [reply]
Ah, you meant that the main.js is not executed by the main process? I just tested on macOs again - worked for me. Not sure some spaces in the path makes some issues here :D

**Comment by Blatman** [comment]
@shalithasuranga  - mmm checked the path and there are no spaces - I have added a wee test app here to show issue. In the debug console the stats log object will show the extensions loaded but not connected. I rejigged the normal resources hierarchy thinking that extensions folder may need to be separate from resources. (built for 10.13 so should be good for higher OS).
Interesting to note that if I set command: in the config to "" I don't get any failure logs at all - I expected it to break.
Thanks

[SampleNeuExtnTest.zip](https://github.com/neutralinojs/neutralinojs/files/7728417/SampleNeuExtnTest.zip)

**Comment by Blatman** [comment]
@shalithasuranga - trying to dig deeper I logged the `// loaded but not connected yet` section of extensions.ts file in client library.
I inserted a 400ms delay before doing a sendMessage but the result was stil a reject with 'not connected' so I can only think that the server side process of loading/connecting is not happening for some reason. Tried shifting the extensions folder around within the app in case it is some sort of path issue but no improvement. I'm not good enough on C++ to attack the server side ....

**Comment by Blatman** [comment, marked as answer]
The answer to this is basically that the sampleextension will work in the CLI because the CLI allows use of node however in an app package node is not available and the config command uses node to start the websocket code. The way round this is to package the extension along with node (in my case using caxa) and alter the config command to run the package ( see #795).
